src/polydatum/middleware.py

```diff
@@ -0,0 +1,177 @@
+    """
+    name = None
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta = None
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+    def __eq__(self, other):
+        return isinstance(other, PathSegment) \
+               and self.name == other.name \
+               and self.meta == other.meta
+
+
+class DalMethodRequest:
+    """
+    Mutable request state. Middleware can modify this.
+    """
+
+    def __init__(
+            self, ctx: DataAccessContext, path: Tuple[PathSegment, ...], args: Tuple[Any, ...], kwargs: Dict
+    ):
+        self.ctx = ctx
+        self.path = path
+        self.args = args
+        self.kwargs = kwargs
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, path: Optional[Tuple[PathSegment]] = None):
+        self.path = path
+        super().__init__(f'Invalid dal method: {".".join([p.name for p in self.path])}')
+
+
+class DeferredDalAttrRequest:
+    """
+    A utility class that defers access to DAL attributes.
+
+    Attribute access to the DAL is deferred so that the DAL can
+    run method middleware before and after calling a method.
+
+    Method middlewares may prevent method's from being called
+    by raising exceptions.
+    """
+
+    # Defining type here allows subclasses to easily provide another class
+    PathSegment = PathSegment
+
+    def __init__(self, handler: Callable, path: Tuple[PathSegment, ...]):
+        """
+        Args:
+            handler (Callable): A callable that handles calling the underlying
+                attribute by wrapping it so that method middleware can run before
+                and after calling a method
+            path (Tuple[PathSegment, ...]): A tuple of PathSegments representing
+                the current attribute path that has been requested.
+        """
+        self._handler = handler
+        self.path = path
+
+    def __getattr__(self, name: str) -> 'DeferredDalAttrRequest':
+        return self.__class__(self._handler, self.path + (self.PathSegment(name=name),))
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+    def __str__(self):
+        return ".".join([p.name for p in self.path])
+
+
+def dal_resolver(ctx, request_path):
+    """
+    This function resolves a dal method call to its underlying
+    service
+    """
+    for (_path, _service_or_method) in resolve(ctx, request_path):
+        if not _service_or_method or not isinstance(_service_or_method, (Callable, Service)):
+            raise DalMethodError(path=_path)
+        service_or_method = _service_or_method
```

Comment:
Bug: It's possible `resolve` does not yield anything. For example, if the `request_path == ()`. In that case, `return service_or_method` will fail because it is not defined.

To fix this, I think having `service_or_method` and `_service_or_method` is not needed. 

```python
service_or_method = None
for (path, service_or_method) in resolve(ctx, request_path):
    if not service_or_method or not isinstance(service_or_method, (Callable, Service)):
        raise DalMethodError(path=path)
return service_or_method
```

Or my preferred grouping:

```python
service_or_method = None
for (path, service_or_method) in resolve(ctx, request_path):
    if not (service_or_method and isinstance(service_or_method, (Callable, Service))):
        raise DalMethodError(path=path)
return service_or_method
```